MREUnityRuntime/MREUnityRuntimeLib/Behaviors/Contexts/BehaviorContextBase.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using MixedRealityExtension.App;
+using MixedRealityExtension.Behaviors.Actions;
+using MixedRealityExtension.Core.Interfaces;
+using MixedRealityExtension.PluginInterfaces.Behaviors;
+using System;
+using System.Linq;
+
+namespace MixedRealityExtension.Behaviors.Contexts
+{
+	public abstract class BehaviorContextBase
+	{
+		private WeakReference<MixedRealityExtensionApp> _appRef;
+		private Guid _attachedActorId;
+		private BehaviorType? _behaviorType;
+
+		internal MixedRealityExtensionApp App
+		{
+			get
+			{
+				if (_appRef.TryGetTarget(out MixedRealityExtensionApp app))
+				{
+					return app;
+				}
+
+				return null;
+			}
+		}
+
+		internal IBehavior Behavior { get; private set; }
+
+		internal BehaviorType BehaviorType
+		{
+			get
+			{
+				_behaviorType = _behaviorType ?? GetBehaviorType();
+				return _behaviorType.Value;
+			}
+		}
+
+		internal BehaviorContextBase()
+		{
+			
+		}
+
+		internal void Initialize(
+			IBehavior behavior,
+			WeakReference<MixedRealityExtensionApp> appRef,
+			IActor attachedActor)
+		{
+			Behavior = behavior;
+			_appRef = appRef;
+			_attachedActorId = attachedActor.Id;
+
+			Behavior.Actor = attachedActor;
+
+			OnInitialized();
+		}
+
+		internal virtual void FixedUpdate()
+		{
+
+		}
+
+		internal virtual void SynchronizeBehavior()
+		{
+
+		}
+
+		internal virtual void CleanUp()
+		{
+			var behavior = Behavior;
+			Behavior = null;
+
+			behavior.CleanUp();
+		}
+
+		protected void RegisterActionHandler(MWActionBase action, string name)
+		{
+			if (Behavior == null) throw new InvalidOperationException("Cannot register action handlers to an uninitialized behavior context.");
```

Comment:
Same concern as below: If this throws, are you certain all codepaths handle it correctly? Will expected reply messages be sent, for example? Wondering if it would be better to log this back to the app instead.

Reply:
See my response for the comment about this in the BehaviorContextFactory.cs file.

---
In reply to: [428863448](https://github.com/microsoft/mixed-reality-extension-unity/pull/160#discussion_r428863448) [](ancestors = 428863448)